packages/tasks_panel/src/misc/e2e-config.ts

```diff
@@ -0,0 +1,220 @@
+    if (project === absPath) {
+      project = ""; // single root
+    }
+    return project;
+  }
+
+  const items: Promise<ProjectInfo | undefined>[] = [];
+  const requestedFolder = Uri.file(wsFolder);
+  const btaExtension: any = extensions.getExtension("SAPOSS.app-studio-toolkit");
+  const basToolkitAPI: BasToolkit = btaExtension?.exports;
+  const workspaceAPI = basToolkitAPI?.workspaceAPI ?? { getProjects: () => Promise.resolve([]) };
+
+  for (const project of await workspaceAPI.getProjects()) {
+    items.push(
+      project.getProjectInfo().then((info) => {
+        if (info) {
+          const workspaceFolder = workspace.getWorkspaceFolder(Uri.file(info.path));
+          if (workspaceFolder?.uri.fsPath.startsWith(requestedFolder.fsPath)) {
+            let style: ProjectTypes | undefined;
+            if (info.type === "com.sap.fe") {
+              style = ProjectTypes.FIORI_FE;
+            } else if (/^com\.sap\.cap(\.java)?$/.test(info.type)) {
+              style = ProjectTypes.CAP;
+            } else if (info.type === "com.sap.hana") {
+              style = ProjectTypes.HANA;
+            }
+            if (style !== undefined) {
+              return {
+                wsFolder,
+                project: asWsRelativePath(info.path),
+                style,
+              };
+            }
+          }
+        }
+      })
+    );
+  }
+  return Promise.all(items).then((items) => compact(items));
+}
+
+export async function addTaskDefinition(wsFolder: string, tasks: TaskDefinition[]): Promise<any> {
+  return updateTasksConfiguration(
+    wsFolder,
+    concat(workspace.getConfiguration("tasks", Uri.file(wsFolder))?.get("tasks") ?? [], tasks)
+  );
+}
+
+export async function isFileExist(uri: Uri): Promise<boolean> {
+  try {
+    return !!(await workspace.fs.stat(uri));
+  } catch (e) {
+    return false;
+  }
+}
+
+export enum LabelType {
+  uniq,
+  sequence,
+}
+
+export async function generateMtaDeployTasks(
+  wsFolder: string,
+  project: string,
+  labelType = LabelType.uniq
+): Promise<TaskDefinition[]> {
+  async function populateCfDetails(): Promise<CfDetails> {
+    try {
+      const targets = await cfGetTargets();
+      if (isEmpty(targets) || (size(targets) === 1 && targets[0].label === DEFAULT_TARGET)) {
+        throw new Error("No CF targets found");
+      }
+      const targetName = find(targets, "isCurrent")?.label;
+      if (!targetName) {
+        throw new Error("No CF current target defined");
+      }
+      /* istanbul ignore next */
+      return {
+        cfTarget: targetName,
+        cfEndpoint: (await cfGetConfigFileField("Target", targetName)) ?? "",
+        cfOrg: (await cfGetConfigFileField("OrganizationFields", targetName))?.Name ?? "",
+        cfSpace: (await cfGetConfigFileField("SpaceFields", targetName))?.Name ?? "",
+      };
+    } catch (e: any) {
+      getLogger().debug(`Can not populate cf target details`, { reason: e.toString() });
```

Comment:
anything can be thrown in JavaScript
so be more careful when `.toString()` on possible Error.
`{ reason: e?.toString() }`

Reply:
`e.toString()` can still `throw` I do not see any handling of the case where
`e` does not have a `toString()` method